dask/array/routines.py

```diff
@@ -211,10 +211,39 @@ def _inner_apply_along_axis(arr,
 def matmul(a, b):
     a = asanyarray(a)
     b = asanyarray(b)
-    try:
-        return a.__matmul__(b)
-    except Exception:
-        return b.__rmatmul__(a)
+
+    if a.ndim == 0 or b.ndim == 0:
+        raise ValueError("`matmul` does not support scalars.")
+
+    a_is_1d = False
+    if a.ndim == 1:
+        a_is_1d = True
+        a = a[None, :]
+
+    b_is_1d = False
+    if b.ndim == 1:
+        b_is_1d = True
+        b = b[:, None]
```

Comment:
I don't know whether `dask.array` cares about such things, but for readability I tend to promote use of `np.newaxis` over `None`.